wagtail/models/__init__.py

```diff
@@ -980,6 +980,11 @@ def get_lock(self):
         if self.locked:
             return BasicLock(self)
 
+        if isinstance(self, WorkflowMixin) and self.current_workflow_task:
```

Comment:
I suppose we could define `get_lock()` in `WorkflowMixin` as well, and use `super().get_lock()`. However, I was thinking of the idea that the `LockableMixin` class being the one who knows all about the default locks we have, and since we already have a `DraftStateMixin` check inside here (even though both are completely independent), might as well have one for `WorkflowMixin` here...

Though we could also argue that we do state that the `WorkflowMixin` requires `LockableMixin`, so we could assume that `LockableMixin` is applied and we could call `super().get_lock()` from `WorkflowMixin`. Maybe that's better. I'll have another look tomorrow

Reply:
If we've gone that way for `DraftStateMixin`, maybe that makes sense then. The pattern of mixins checking which other mixins are applied feels slightly awkward in places - here I think the method + `super()` in `WorkflowMixin` would work for the logic, did it not work neatly for `DraftStateMixin` then?

Reply:
With `DraftStateMixin`, locking is completely optional (and was introduced later when we added `LockableMixin` support for snippets), thus I felt like it shouldn't need to know anything about a `get_lock()` method. I initially thought it would be more readable if the lock is always returned from `LockableMixin`, so we know what the order of precedence is.

However, now that we say `WorkflowMixin` requires `LockableMixin`, it does feel a bit weird for `LockableMixin` to know about `WorkflowMixin`, so I guess using `super()` here works better.

Unfortunately, if we'd like to do the same for `DraftStateMixin`, we'll have to add another check to ensure that the MRO for `DraftStateMixin` comes first before `LockableMixin` (or, we add a `super().get_lock()` in `LockableMixin` itself, which seems a bit silly). We didn't add any explicit checks to ensure this when we introduced `LockableMixin` for snippets because both mixins can be used independently. If we want to do this, now would be the best time before the RC is out... But I feel like it's out of the scope of this PR?

In the interest of time, I'd say we make `WorkflowMixin` use `super().get_lock()`, but leave the check for `DraftStateMixin`. This way, we could make `LockableMixin` not rely on `WorkflowMixin`, but keep the "canonical" order of the mixins exactly as the order they were introduced: `WorkflowMixin`, `LockableMixin`, `DraftStateMixin`, `RevisionMixin`.

Reply:
Yep, I agree with that. That said - is there anything intrinsic to `LockableMixin` required by `WorkflowMixin` having a `get_lock` method? It doesn't use any of the fields on that mixin, for example. So in theory it could be independent, but would rely on views/other areas checking for a `get_lock` method rather than `LockableMixin`,

For now, yeah, let's go with `WorkflowMixin` defines its own `get_lock` method and calls `super()` - I think that makes most sense

Reply:
> Yep, I agree with that. That said - is there anything intrinsic to `LockableMixin` required by `WorkflowMixin` having a `get_lock` method? It doesn't use any of the fields on that mixin, for example. So in theory it could be independent, but would rely on views/other areas checking for a `get_lock` method rather than `LockableMixin`,

Nope, you're right there isn't. It's just that I'm not sure if that's something we want to support officially. It's much easier to maintain if we say that you need `LockableMixin` when using `WorkflowMixin` from the get-go instead of having to guard against all possible combinations. Though I'll see if I can make the workflow code safer without `LockableMixin` applied, and see if this is something we can support.